src/hotspot/share/gc/shenandoah/shenandoahGenerationalHeap.cpp

```diff
@@ -96,3 +116,153 @@ void ShenandoahGenerationalHeap::stop() {
+           "Unaffiliated available must be less than total available");
+    size_t old_fragmented_available =
+            old_available - old_generation()->free_unaffiliated_regions() * region_size_bytes;
+    reserve_for_mixed = max_evac_need + old_fragmented_available;
+    if (reserve_for_mixed > max_old_reserve) {
+      reserve_for_mixed = max_old_reserve;
+    }
+  }
+
+  // Decide how much space we should reserve for promotions from young
+  size_t reserve_for_promo = 0;
+  const size_t promo_load = old_generation()->get_promotion_potential();
+  const bool doing_promotions = promo_load > 0;
+  if (doing_promotions) {
+    // We're promoting and have a bound on the maximum amount that can be promoted
+    assert(max_old_reserve >= reserve_for_mixed, "Sanity");
+    const size_t available_for_promotions = max_old_reserve - reserve_for_mixed;
+    reserve_for_promo = MIN2((size_t)(promo_load * ShenandoahPromoEvacWaste), available_for_promotions);
+  }
+
+  // This is the total old we want to ideally reserve
+  const size_t old_reserve = reserve_for_mixed + reserve_for_promo;
+  assert(old_reserve <= max_old_reserve, "cannot reserve more than max for old evacuations");
+
+  // We now check if the old generation is running a surplus or a deficit.
+  size_t old_region_deficit = 0;
+  size_t old_region_surplus = 0;
+
+  const size_t max_old_available = old_generation()->available() + old_cset_regions * region_size_bytes;
+  if (max_old_available >= old_reserve) {
+    // We are running a surplus, so the old region surplus can go to young
+    const size_t old_surplus = max_old_available - old_reserve;
+    old_region_surplus = old_surplus / region_size_bytes;
+    const size_t unaffiliated_old_regions = old_generation()->free_unaffiliated_regions() + old_cset_regions;
+    old_region_surplus = MIN2(old_region_surplus, unaffiliated_old_regions);
+  } else {
+    // We are running a deficit which we'd like to fill from young.
+    // Ignore that this will directly impact young_generation()->max_capacity(),
+    // indirectly impacting young_reserve and old_reserve.  These computations are conservative.
+    const size_t old_need = old_reserve - max_old_available;
+    // The old region deficit (rounded up) will come from young
+    old_region_deficit = (old_need + region_size_bytes - 1) / region_size_bytes;
+
+    // Round down the regions we can transfer from young to old. If we're running short
+    // on young-gen memory, we restrict the xfer. Old-gen collection activities will be
+    // curtailed if the budget is restricted.
+    const size_t max_old_region_xfer = old_xfer_limit / region_size_bytes;
+    old_region_deficit = MIN2(old_region_deficit, max_old_region_xfer);
+  }
+  assert(old_region_deficit == 0 || old_region_surplus == 0, "Only surplus or deficit, never both");
+
+  old_generation()->set_region_surplus(old_region_surplus);
+  old_generation()->set_region_deficit(old_region_deficit);
```

Comment:
We could just do a single signed quantity here. Not sure if the computation of the surplus/deficit could be simplified. Likely not.